File: iteration2/src/models/Transcript.java
package iteration2.src.models;
import iteration2.src.enums.CourseResult;
import iteration2.src.enums.CourseType;

import java.util.*;

import iteration2.src.utils.DatabaseManager;

public class Transcript{

    private List <CourseGrade> takenCourses;

    public Transcript() {
    }

    public Transcript(List <CourseGrade> takenCourses){
        this.takenCourses = takenCourses;
    }

    public void addCourse(Course course){
        takenCourses.add(new CourseGrade(course, null, CourseResult.ACTIVE));
    }


    
    public List<Course> acquirePassedCourses() {

        List<Course> passedCourses = new ArrayList<Course>();
        for (CourseGrade course : takenCourses) {
            if (course.getCourseResult() == CourseResult.PASSED) {
                passedCourses.add(course.getCourse());
            }
        }
        return passedCourses;
    }
    
    public double calculateCompletedCredits(){
        double total = 0.0;
        for(CourseGrade course : takenCourses){
            if(course.getCourseResult() == CourseResult.PASSED)
                total += course.getCourse().getCourseCredit();
        }
        return total;
    }

    public double calculateGPA(){
        double totalPoint = 0.0;
        double totalCredit = 0.0;
        for(CourseGrade course : takenCourses){
            totalPoint += course.getCourse().getCourseCredit()*course.convertLetterGradeToScore();
            totalCredit += course.getCourse().getCourseCredit();
        }

        return totalPoint/totalCredit;
    }

    public List<CourseGrade> getTakenCourses() {
        return takenCourses;
    }

    public void setTakenCourses(List<CourseGrade> takenCourses) {
        this.takenCourses = takenCourses;
    }

    
    //This method checks if the student has taken all the courses that are required for graduation
    //Which is at least 165 credits exluding nte, te, ue, fe and isg
    public boolean checkEngineeringProjectAvailability() {
        HashMap<Integer,String> constraint = DatabaseManager.getInstance().getConstraints();

        double totalCreditsForEngineeringProject = 0;

        //Those course types are not count for engineering project
        CourseType restrictedCourseTypes[] = {
            CourseType.NONTECHNICAL_ELECTIVE,
            CourseType.UNIVERSITY_ELECTIVE,
            CourseType.TECHNICAL_ELECTIVE,
            CourseType.FACULTY_ELECTIVE,
        };

        //Those courses are not count for engineering project
        String restrictedCourses[] = {"ISG121", "ISG122"};
        

        for(int i = 0; i < takenCourses.size(); i++){

            if(takenCourses.get(i).getCourseResult() == CourseResult.PASSED && 
                Arrays.asList(restrictedCourseTypes).contains(takenCourses.get(i).getCourse().getCourseType()) == false &&
                Arrays.asList(restrictedCourses).contains(takenCourses.get(i).getCourse().getCourseCode()) == false) {


                totalCreditsForEngineeringProject += takenCourses.get(i).getCourse().getCourseCredit();

            }
        }

        if (totalCreditsForEngineeringProject >= Integer.parseInt(constraint.get(3))) {
            return true;
        } else {
            return false;
        }
    }

    

    
    
}

File: iteration3/src/models/Transcript.py
from iteration3.src.models.CourseGrade import CourseGrade
from iteration3.src.enums.CourseResult import CourseResult
from iteration3.src.utils.DatabaseManager import DatabaseManager
from iteration3.src.enums.CourseType import CourseType


class Transcript(object):
    
    def __init__(self, takenCourses):
        self._takenCourses = takenCourses

    @classmethod
    def dictToObject(cls, dict):
        return cls(
            [CourseGrade.dictToObject(courseGrade) for courseGrade in dict['takenCourses']]
        )
    
    def addTakenCourse(self, course):
        self._takenCourses.append( CourseGrade(course, None, CourseResult.ACTIVE) )

    def acquirePassedCourses(self):
        passedCourses = list()
        for course in self._takenCourses:
            if course.getCourseResult() == CourseResult.PASSED:
                passedCourses.append(course)
        
        return passedCourses
    
    def calculateCompletedCredits(self):
        total = 0.0
        for course in self._takenCourses:
            if course.getCourseResult() == CourseResult.PASSED:
                total += course.getCourse().getCourseCredit()
        
        return total
    
    def calculateGPA(self):
        totalPoint = 0.0
        totalCredit = 0.0

        for courseGrade in self._takenCourses:
            if courseGrade.getCourseResult() != CourseResult.PASSED:
                continue
            totalPoint += courseGrade.getCourse().getCourseCredit() * courseGrade.convertLetterGradeToScore()
            totalCredit += courseGrade.getCourse().getCourseCredit()
        try:
            return totalPoint / totalCredit
        except ZeroDivisionError:
            return 0
    
    def checkEngineeringProjectAvailability(self):
        totalCreditsForEngineeringProject = 0.0

        constraintsMap = DatabaseManager.getInstance().getConstraints()

        restricedCourses = ["ISG121", "ISG122"]
        restricedCourseTypes = [CourseType.NONTECHNICAL_ELECTIVE, CourseType.UNIVERSITY_ELECTIVE,
                                    CourseType.TECHNICAL_ELECTIVE,CourseType.FACULTY_ELECTIVE,]
        
        for takenCourse in self._takenCourses:
           if (takenCourse.getCourseResult() == CourseResult.PASSED and 
               takenCourse.getCourse().getCourseType() not in restricedCourseTypes and
                takenCourse.getCourse().getCourseCode() not in restricedCourses ):
               totalCreditsForEngineeringProject += takenCourse.getCourse().getCourseCredit()

        if totalCreditsForEngineeringProject >= int(constraintsMap[3]):
            return True
        else:
            return False
        
    def getTakenCourses(self):
        return self._takenCourses